### PR TITLE
Prevent client skew

### PR DESCRIPTION
**Pull request by deads2k**
Adds an option to prevent skewed oc clients from performing PUT or POST (update or create) actions on the API server.  This can prevent weirdness where fields are dropped by old oc clients.  The default is to avoid restricting any client.  Any client not strictly conforming to `DefaultOpenShiftUserAgent` is allowed.

Also fixed here: update the server and client to always identify as openshift, not as kubernetes.  Our client is not the same as kube and the commits didn't really line up due to patches anyway.

@fabianofranz @jwforres since you asked.
@liggitt hold your nose, I don't think its ever going to smell nice.

Buzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1306590

### CODE REVIEW

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/helper.go [hunk 1]
```diff
@@ -580,7 +580,7 @@ func DefaultKubernetesUserAgent() string {
 	version := version.Get().GitVersion
 	seg := strings.SplitN(version, "-", 2)
 	version = seg[0]
-	return fmt.Sprintf("%s/%s (%s/%s) kubernetes/%s", path.Base(os.Args[0]), version, gruntime.GOOS, gruntime.GOARCH, commit)
+	return fmt.Sprintf("%s/%s (%s/%s) openshift/%s", path.Base(os.Args[0]), version, gruntime.GOOS, gruntime.GOARCH, commit)
```
Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/helper.go]: origin vs openshift vs aep?

File: hack/common.sh [hunk 2]
```diff
@@ -711,11 +711,11 @@ os::build::ldflags() {
...
   if [[ -z $(echo "${GO_VERSION[2]}" | grep -E 'go1.5') ]]; then
-    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitCommit" "${KUBE_GIT_COMMIT}")
-    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitVersion" "${KUBE_GIT_VERSION}")
+    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitCommit" "${OS_GIT_COMMIT}")
+    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitVersion" "${OS_GIT_VERSION}")
   else
-    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitCommit=${KUBE_GIT_COMMIT}")
-    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitVersion=${KUBE_GIT_VERSION}")
+    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitCommit=${OS_GIT_COMMIT}")
+    ldflags+=(-X "k8s.io/kubernetes/pkg/version.gitVersion=${OS_GIT_VERSION}")
```
Comment (liggitt) [inline hack/common.sh]: what does this do to `oc version` and `openshift version`? those used to spit out origin and k8s versions

File: pkg/cmd/server/origin/handlers.go [hunk 3]
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
...
+		// that value is passed for a 1.1.3 client when accessing kube resources
+		if strings.Contains(userAgent, "kubernetes/") {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// user agent is: binary/versionNumber (operating system) openshift/commit
+		// for example: oc/v1.1.3 (linux/amd64) openshift/9b52c41
+		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
+		// stop stomping things they don't recognize
```
Comment (liggitt) [inline pkg/cmd/server/origin/handlers.go]: piecemill parsing is super hard to follow... I can't convince you to use a regex here can I?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/helper.go (hunk 1 above)
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/client/unversioned/helper.go]: > origin vs openshift vs aep?

"openshift" is hardcoded in https://github.com/openshift/origin/blob/master/pkg/client/client.go#L315 which this now matches.

File: hack/common.sh (hunk 2 above)
Comment (deads2k) [inline hack/common.sh]: > what does this do to oc version and openshift version? those used to spit out origin and k8s versions

Messes them up.  Guess I'd rather pull that entirely.  Objections?

Comment (liggitt) [inline hack/common.sh]: I vaguely liked having the info available, but I'll let @smarterclayton weigh in

File: pkg/cmd/server/origin/handlers.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/server/origin/handlers.go]: > piecemill parsing is super hard to follow... I can't convince you to use a regex here can I?

`oc/v(\w+) \(.+/.+\) openshift/(\w{7}0`?  can't say that I think its prettier, but if you like it better....

File: hack/common.sh (hunk 2 above)
Comment (deads2k) [inline hack/common.sh]: > I vaguely liked having the info available, but I'll let @smarterclayton weigh in

Do you have any good advice for making sure that we never accidentally start using after rebases?  Using it in places like command line clients (but only for some resources), `MatchesServerVersion` (but only sometimes), `/version` endpoint?  Mechanics-wise it is painful to keep and of dubious value.  When have you used it?  Was it more useful than knowing the origin level?

Comment (smarterclayton) [inline hack/common.sh]: Users and clients want to know what version of Kube we are. We have to
keep it.  We want to get better at describing exactly what version we
include.  We can't really change /version since that is a Kube API

File: pkg/cmd/server/origin/handlers.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/origin/handlers.go]: I find it easier to see the structure of what is being matched with something like `oc/v(\w+) .* openshift/(\w{7})`, but maybe that's just me

File: hack/common.sh (hunk 2 above)
Comment (deads2k) [inline hack/common.sh]: > Users and clients want to know what version of Kube we are. We have to
> keep it.  We want to get better at describing exactly what version we
> include.  We can't really change /version since that is a Kube API

Ok, then we have multiple problems:
 1. commit sha points to our fork, not to kube.  This means that a kubectl client will *never* match our server
 1. `/version` returns kube information. This means that an oc client will *never* match our server.

Conclusion: no client ever matches our server.
Conclusion: no one has ever used this because it doesn't work.

Solutions?
 1. respond to `/version` differently based on "User-Agent".
 1. create a second endpoint: `/openshift-versions` that provides enough structure to make a determination
 1. find a way to locate the actual kube tags and commit shas (kind of required)

File: pkg/cmd/server/origin/handlers.go (hunk 3 above)
Comment (fabianofranz) [inline pkg/cmd/server/origin/handlers.go]: +1 regex

File: hack/common.sh (hunk 2 above)
Comment (smarterclayton) [inline hack/common.sh]: I guess I'm ok with /versions reporting our commit.  kubectl will
never match our commit, I mostly care about oc.  I actually don't care
about this feature much - I don't see it *ever* being useful for
anyone in real world.

Comment (smarterclayton) [inline hack/common.sh]: I cannot think of a scenario where these are going to match except in RPM installs, and maybe for like 2 days on Online.

File: pkg/cmd/server/origin/handlers.go [hunk 4]
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
...
 }
 
+// versionSkewFilter adds a filter that may deny PUT an POST requests since they stomp entire resource objects from skewed
+// oc clients, since we know that those clients will strip unknown fields which can lead to unexpected outcomes
+func (c *MasterConfig) versionSkewFilter(binaryInfo version.Info, handler http.Handler) http.Handler {
+	skewedClientPolicy := c.Options.PolicyConfig.SkewedClientPolicy
+	if skewedClientPolicy == configapi.AllowAll {
+		return handler
+	}
+	seg := strings.SplitN(binaryInfo.GitVersion, "-", 2)
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/handlers.go]: If you need this split it should be done in the build, not here.  This is fragile.

File: pkg/cmd/server/origin/handlers.go [hunk 5]
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
...
+	serverVersionNumber := seg[0][1:]
+	serverCommit := binaryInfo.GitCommit
+	if len(serverCommit) > 7 {
+		serverCommit = serverCommit[:7]
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		if req.Method != "PUT" && req.Method != "POST" {
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/handlers.go]: And PATCH.

File: pkg/cmd/server/api/types.go [hunk 6]
```diff
@@ -317,8 +317,24 @@ type PolicyConfig struct {
 
 	// OpenShiftInfrastructureNamespace is the namespace where OpenShift infrastructure resources live (like controller service accounts)
 	OpenShiftInfrastructureNamespace string
+
+	// SkewedClientPolicy controls how API calls from *voluntarily* identifying clients will be handled.  THIS DOES NOT DEFEND AGAINST MALICIOUS CLIENTS!
+	SkewedClientPolicy SkewedClientPolicy
```
Comment (smarterclayton) [inline pkg/cmd/server/api/types.go]: LegacyClientPolicy.   Skewed doesn't have the same meaning to users.

File: pkg/cmd/server/origin/handlers.go [hunk 7]
```diff
@@ -168,6 +176,98 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
...
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		// user agent is: binary/versionNumber (operating system) openshift/commit
+		// for example: oc/v1.1.3 (linux/amd64) openshift/9b52c41
+		// only match on oc.  This is very restrictive check.  We want to allow by default here.  Clients effectively "opt-in" to
+		// stop stomping things they don't recognize
+		tokens := strings.SplitN(userAgent, "oc/v", 2)
+		if len(tokens) != 2 {
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/handlers.go]: "oc" is not the only client we will need to do this for.  So we should be prepared to expand to multiple client prefixes and types.

File: pkg/cmd/server/origin/handlers.go (hunk 5 above)
Comment (deads2k) [inline pkg/cmd/server/origin/handlers.go]: Not patch.  Patches are non-encoded json that avoids any field stripping logic.  That stripping logic is the only reason to block skewed usage.

File: pkg/cmd/server/origin/handlers.go (hunk 4 above)
Comment (deads2k) [inline pkg/cmd/server/origin/handlers.go]: This is done upstream to build the user agent string.  Change it there too?  It would be a lot easier to do if the type wasn't unversioned.

So, same logic in both spots, new field duplicating information, or replumb the logic through both sides discrete fields?

File: pkg/cmd/server/origin/handlers.go (hunk 7 above)
Comment (deads2k) [inline pkg/cmd/server/origin/handlers.go]: There's no reason to believe they'd submit the same format of user agent.  The code can be reworked if and when that is required, probably with priority regex matching, it's not particularly dense or layered.

**Comment by deads2k** [comment]
Updated.  `/version` still responds as though it were a strange level of a kube API server.  When accessing kube resources, our client still masquerades as a `kubectl` so we can keep tricking kube API servers (if they care).  When accessing origin resources, our client identifies as openshift.

The server-side filtering now compares provided kube levels against our embedded kube level and provided openshift levels against our openshift level.  I've added a twiddle knob to control the HTTP verbs the skew filter applies to in response to user requests for `GET` prevention.

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
@pweil- ptal.

File: pkg/cmd/server/origin/handlers.go [hunk 8]
```diff
@@ -168,6 +178,91 @@ func namespacingFilter(handler http.Handler, contextMapper kapi.RequestContextMa
...
+// 5. oadm kube resources: oadm/v1.2.0 (linux/amd64) kubernetes/bc4550d
+// 6. oadm openshift resources: oadm/v1.1.3 (linux/amd64) openshift/b348c2f
+// 7. openshift cli kube resources: openshift/v1.2.0 (linux/amd64) kubernetes/bc4550d
+// 8. openshift cli openshift resources: openshift/v1.1.3 (linux/amd64) openshift/b348c2f
+var (
+	kubeStyleUserAgent      = regexp.MustCompile(`\w+/v([\w\.]+) \(.+/.+\) kubernetes/\w{7}`)
+	openshiftStyleUserAgent = regexp.MustCompile(`\w+/v([\w\.]+) \(.+/.+\) openshift/\w{7}`)
+)
+
+// versionSkewFilter adds a filter that may deny PUT an POST requests since they stomp entire resource objects from skewed
```
Comment (pweil-) [inline pkg/cmd/server/origin/handlers.go]: typo: `PUT an POST`

File: pkg/cmd/server/origin/handlers_test.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	config := MasterConfig{}
+	config.Options.PolicyConfig.LegacyClientPolicyConfig.LegacyClientPolicy = configapi.AllowAll
+	config.Options.PolicyConfig.LegacyClientPolicyConfig.RestrictedHTTPVerbs = verbs
+	server := httptest.NewServer(config.versionSkewFilter(openshiftServerVersion, kubeServerVersion, doNothingHandler))
+	defer server.Close()
+
+	testCases := []versionSkewTestCase{
+		{
+			name:    "missing",
+			methods: verbs,
```
Comment (pweil-) [inline pkg/cmd/server/origin/handlers_test.go]: The `missing` cases in the tests need to have a slice with an empty agent to have something to `range` over in order to invoke the tests.

**Comment by deads2k** [comment]
comments addressed.  Plan to merge on green test unless someone has something else.

**Comment by deads2k** [comment]
[merge]

File: pkg/cmd/server/origin/handlers.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/handlers.go]: No, skewed usage checks are not only useful for stripping logic, they are
primarily used by admins to force clients to upgrade because the logic the
client is using in the patch is faulty.  The *goal* of skew checks is to
deny clients.  *sometimes* the reason a client is denied is because it
doesn't follow correct logic.

**Comment by smarterclayton** [comment]
Undid merge pending my comments.

File: pkg/cmd/server/origin/handlers.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/handlers.go]: It's fine to have logic in both spots for now, we just need to have an
upstream issue to avoid having to parse.  No one ever spent enough time on
this, really.

File: pkg/cmd/server/origin/handlers.go (hunk 7 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/handlers.go]: As mentioned above the point of a skew client check is to give admins a
tool to prevent problems they hit.  Oc isn't the only client - the fact
that we are only worried about it right now is just because it's the first
example, but the actual feature we want is "deny clients based on user
agent".  The selective stripping checks are an optimization for a
particular problem (to be selective), but the more general "deny clients
that report these user agents and give them a clean message" is the feature
we want to offer to admins.

File: pkg/cmd/server/origin/handlers.go (hunk 5 above)
Comment (deads2k) [inline pkg/cmd/server/origin/handlers.go]: > No, skewed usage checks are not only useful for stripping logic, they are
> primarily used by admins to force clients to upgrade because the logic the
> client is using in the patch is faulty.  The *goal* of skew checks is to
> deny clients.  *sometimes* the reason a client is denied is because it
> doesn't follow correct logic.

This is an outdated diff because you can now choose which verbs you want to apply the check to, so its possible to operate against PATCH.

I still maintain that PATCH is safe and that only really, really, really borked clients will generate patches that valid, but have unintended behavior.

**Comment by deads2k** [comment]
> The selective stripping checks are an optimization for a
> particular problem (to be selective), but the more general "deny clients
> that report these user agents and give them a clean message" is the feature
> we want to offer to admins.

Do you want to build that full logic now?  I find it difficult to believe that its flexibility we need now.  The code could be extended to take things like user-regexes and display strings for matching purposes, but that's a significant amount of creep for what seems like a really rare case.

**Comment by smarterclayton** [comment]
The customers are actually asking for that feature, not this.

**Comment by deads2k** [comment]
One comment seems spurious (PATCH can be added via config), one is a request for a followup issue: https://github.com/openshift/origin/issues/7838.

**Comment by smarterclayton** [comment]
Not sure what you mean by spurious, PATCH should default on (unless you've
now changed it so the list is always from config, in which case I'm ok)

**Comment by deads2k** [comment]
> Not sure what you mean by spurious, PATCH should default on (unless you've
> now changed it so the list is always from config, in which case I'm ok)

The list is always from the config.

**Comment by smarterclayton** [comment]
Lgtm, reapply as necessary

**Comment by deads2k** [comment]
[merge]
